File: solutions/day_02/advent2.1.rb
#!/usr/bin/env ruby

def input
  @input ||= STDIN.read.split("\n").map{|row| row.split("\t").map(&:to_i)}
end

def calculate_checksum
  input.map do |row|
    row.max - row.min
  end.reduce :+
end

if __FILE__ == $0
  p calculate_checksum
end

File: solutions/day_02/advent2.1.exs
#!/usr/bin/env elixir

defmodule Checksum do
  def fetch_input do
    IO.read(:stdio, :all)
    |> String.split("\n")
    |> Enum.map(fn (row)->
      String.split(row, "\t")
      |> Enum.map(&Integer.parse/1)
      |> Enum.map(fn ({x,_}) ->
        x
      end)
    end)
  end

  def calculate_checksum(input) do
    Enum.map(input, fn (row) ->
      Enum.max(row) - Enum.min(row)
    end)
    |> Enum.sum 
  end
end

Checksum.fetch_input
|> Checksum.calculate_checksum
|> IO.inspect 

